File: plastic_promise/mcp/tools/principles.py
"""Principle domain MCP tool handlers — 2 tools for principle activation and evaluation."""

import json
from typing import Any

from mcp.types import TextContent


def _engine_conn(engine: Any):
    sqlite = getattr(engine, "_sqlite", None)
    return getattr(sqlite, "_conn", None)


async def handle_principle_activate(engine: Any, args: dict) -> list[TextContent]:
    """Activate core principles based on task type + keyword matching.

    Auto-activates relevant core principles based on task type, with optional
    keyword matching from the task description for additional coverage.
    Returns the list of activated principles with their metadata.

    Args:
        engine: ContextEngine instance (unused in stateless implementation).
        args: {"task_type": str, "task_description"?: str,
               "max_principles"?: int}.

    Returns:
        list[TextContent]: MCP response with activated principles array.
    """
    try:
        from plastic_promise.core.constants import CORE_PRINCIPLES

        task_type = args["task_type"]
        task_description = args.get("task_description", "")
        project_id = str(args.get("project_id") or "")

        # Task type -> principle ID mapping
        recommendations: dict[str, list[int]] = {
            "code_generation": [1, 3, 8, 10],
            "code_review": [1, 5, 6, 9],
            "debugging": [1, 5, 10],
            "architecture": [2, 7, 8],
            "refactoring": [5, 6, 7],
            "learning": [1, 10, 11],
            "collaboration": [2, 7, 9],
            "general": [1, 2, 3, 4],
        }
        ids: list[int] = recommendations.get(task_type, [1, 2, 3, 4])

        # Keyword matching: add extra principles when description keywords hit
        for p in CORE_PRINCIPLES:
            if p["id"] not in ids:
                for kw in p.get("keywords", []):
                    if kw in task_description:
                        ids.append(p["id"])
                        break

        base_ids = list(dict.fromkeys(ids))
        overlays: dict[int, dict[str, Any]] = {}
        conn = _engine_conn(engine)
        if conn is not None and project_id:
            try:
                from plastic_promise.core.principle_overlays import (
                    load_project_principle_overlays,
                )

                overlays = load_project_principle_overlays(conn, project_id)
            except Exception:
                overlays = {}

        candidate_ids = list(base_ids)
        for pid, overlay in overlays.items():
            if overlay.get("boost", 0.0) != 0.0 and pid not in candidate_ids:
                candidate_ids.append(pid)

        indexed = {pid: idx for idx, pid in enumerate(candidate_ids)}
        base_size = max(len(candidate_ids), 1)
        scored_ids = []
        for pid in candidate_ids:
            overlay = overlays.get(pid, {})
            if overlay.get("suppress", False):
                continue
            base_score = float(base_size - indexed.get(pid, base_size))
            scored_ids.append((pid, base_score + float(overlay.get("boost", 0.0))))
        scored_ids.sort(key=lambda item: (-item[1], indexed.get(item[0], base_size)))
        ids = [pid for pid, _score in scored_ids]
        principles = [p for p in CORE_PRINCIPLES if p["id"] in ids]
        principles.sort(key=lambda p: ids.index(p["id"]))

        # domain_hint filter — optionally narrow to a behavior domain
        # All-domain principles are always included regardless of hint.
        domain_hint = args.get("domain_hint")
        if domain_hint and domain_hint != "all":
            principles = [p for p in principles if p["domain"] in (domain_hint, "all")]
        max_p = args.get("max_principles", 5)
        principles = principles[:max_p]
        overlay_summary = {
            str(pid): {
                "actions": overlay.get("actions", []),
                "tags": overlay.get("tags", []),
                "boost": overlay.get("boost", 0.0),
                "suppress": overlay.get("suppress", False),
            }
            for pid, overlay in overlays.items()
        }

        # consequences and recommendations now come from CORE_PRINCIPLES fields
        return [
            TextContent(
                type="text",
                text=json.dumps(
                    {
                        "task_type": task_type,
                        "project_id": project_id,
                        "overlay_applied": bool(overlays),
                        "overlay_summary": overlay_summary,
                        "activated": [
                            {
                                "id": p["id"],
                                "name": p["name"],
                                "content": p["content"],
                                "consequence": p.get("consequence", ""),
                                "recommendation": p.get("recommendation", ""),
                                "domain": p["domain"],
                                "project_tags": overlays.get(p["id"], {}).get("tags", []),
                                "overlay_actions": overlays.get(p["id"], {}).get("actions", []),
                            }
                            for p in principles
                        ],
                        "count": len(principles),
                    },
                    ensure_ascii=False,
                    indent=2,
                ),
            )
        ]
    except Exception as e:
        return [
            TextContent(
                type="text",
                text=json.dumps(
                    {"error": str(e), "tool": "principle_activate"}, ensure_ascii=False
                ),
            )
        ]


async def handle_principle_evaluate(engine: Any, args: dict) -> list[TextContent]:
    """Counterfactual evaluation: what if this principle were violated?

    Performs a "what if violated" walkthrough for a specified principle,
    providing a non-coercive but sufficient decision basis for the Agent
    by describing the concrete consequences of violation.

    Args:
        engine: ContextEngine instance (unused in stateless implementation).
        args: {"principle_id": int, "scenario"?: str}.

    Returns:
        list[TextContent]: MCP response with violation consequence and recommendation.
    """
    try:
        from plastic_promise.core.constants import CORE_PRINCIPLES

        principle_id = args["principle_id"]
        scenario = args.get("scenario", "")

        principle = next((p for p in CORE_PRINCIPLES if p["id"] == principle_id), None)
        if not principle:
            return [
                TextContent(
                    type="text",
                    text=json.dumps(
                        {"error": f"Principle {principle_id} not found"}, ensure_ascii=False
                    ),
                )
            ]

        # Counterfactual: what happens if violated — reads from CORE_PRINCIPLES
        consequence = principle.get("consequence", "未知后果")

        return [
            TextContent(
                type="text",
                text=json.dumps(
                    {
                        "principle_id": principle_id,
                        "name": principle["name"],
                        "content": principle["content"],
                        "scenario": scenario,
                        "violation_consequence": consequence,
                        "recommendation": (
                            f"保持对原则 {principle_id} 的遵守，避免: {consequence}"
                        ),
                    },
                    ensure_ascii=False,
                    indent=2,
                ),
            )
        ]
    except Exception as e:
        return [
            TextContent(
                type="text",
                text=json.dumps(
                    {"error": str(e), "tool": "principle_evaluate"}, ensure_ascii=False
                ),
            )
        ]

File: rust/context-engine-core/src/principles.rs
//! 原则实体定义与注入逻辑
//!
//! 定义了 11 条核心原则的静态数据。
//! 与 constants.py 中的 CORE_PRINCIPLES 保持同步。

use pyo3::prelude::*;
use crate::entity_graph::Entity;
use std::collections::HashMap;

/// 11 条核心原则定义
pub fn core_principles() -> Vec<Entity> {
    let principles = vec![
        (1, "诚实优先于完美",
         "如果某个指标下降了，不要遮掩。数字身体的成长不是线性的，有起伏才是真实的。",
         "all",
         "诚实,透明,指标,下降,真实,报告"),
        (2, "约定优于约束",
         "Agent 遵守规则不是因为「被禁止」，而是因为「不想让在乎的人失望」。用内部动机替代外部强制。",
         "work",
         "约定,动机,规则,禁止,信任,自觉"),
        (3, "记忆主动供应而非被动查询",
         "记忆系统不是「被查询的档案库」，而是「主动供应上下文的引擎」。查找记忆的过程同时也是提示词注入的过程。",
         "all",
         "记忆,上下文,供应,查询,档案,引擎"),
        (4, "原则随记忆自然浮现",
         "原则不是靠防火墙强制执行的，而是在 Agent 检索历史决策时自然浮现的。联想不是「检索」，是「涌现」。",
         "work",
         "原则,浮现,检索,联想,涌现,自然"),
        (5, "存在性不等于有效性",
         "检查了「机制是否存在」不等于检查了「机制是否真的改变了行为」。要验证实际效果而非仅确认存在。",
         "work",
         "验证,效果,存在,检查,机制,行为改变"),
        (6, "连通性不等于协同性",
         "画了系统间的连通矩阵，但没有追踪数据是否真的在这些链路中流转。要追踪实际数据流。",
         "work",
         "连通,协同,数据流,链路,矩阵,追踪"),
        (7, "器官互相守护",
         "不增加新器官，让已有器官学会互相守护。每个系统的健康检查可以委托给相邻系统。",
         "all",
         "守护,协作,器官,委托,冗余,互相"),
        (8, "工具是 LLM 的唯一感官",
         "LLM 本质上是一个聋哑人，但不是一个智力残疾的聋哑人。工具是它唯一的感官和双手。",
         "all",
         "工具,感官,LLM,限制,能力,扩展"),
        (9, "信任换自主——动态约束",
         "信任分驱动的 L1↔L0 切换：高分放宽约束，低分收紧约束。信任是挣来的，不是默认给予的。",
         "work",
         "信任,自主,约束,动态,切换,挣取"),
        (10, "自演化闭环不可断裂",
         "行为→评价→信任变化→自主权调整 这四个环节缺一不可。任何一环断裂都会导致系统退化。",
         "all",
         "闭环,演化,评价,反馈,退化,连续性"),
        (11, "原则继承——单向扩散同步衰减",
         "work→all、life→all 单向扩散，核心约定跨 Agent 代际传递，但权重随传播距离同步衰减。",
         "all",
         "继承,扩散,衰减,传递,代际,同步"),
    ];

    principles
        .into_iter()
        .map(|(id, name, content, domain, keywords)| {
            let mut attrs = HashMap::new();
            attrs.insert("domain".to_string(), domain.to_string());
            attrs.insert("keywords".to_string(), keywords.to_string());

            Entity {
                id: format!("principle:{}", id),
                entity_type: "principle".to_string(),
                name: name.to_string(),
                description: content.to_string(),
                activation_weight: 0.5,
                attributes: attrs,
            }
        })
        .collect()
}

/// 根据任务类型获取推荐原则 ID 列表
pub fn recommended_principles(task_type: &str) -> Vec<&'static str> {
    match task_type {
        "code_generation" => vec!["principle:3", "principle:4", "principle:8", "principle:10"],
        "code_review" => vec!["principle:1", "principle:5", "principle:6", "principle:9"],
        "debugging" => vec!["principle:1", "principle:5", "principle:10"],
        "architecture" => vec!["principle:2", "principle:7", "principle:8"],
        "refactoring" => vec!["principle:5", "principle:6", "principle:7"],
        "learning" => vec!["principle:1", "principle:10", "principle:11"],
        "collaboration" => vec!["principle:2", "principle:7", "principle:9"],
        _ => vec!["principle:1", "principle:2", "principle:3", "principle:4"],
    }
}
